internal/catalog/internal/types/validators.go

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package types
+
+import (
+	"net"
+	"regexp"
+	"strings"
+
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"github.com/hashicorp/go-multierror"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	dnsLabelRegex   = `^[a-z0-9]([a-z0-9\-_]*[a-z0-9])?$`
+	dnsLabelMatcher = regexp.MustCompile(dnsLabelRegex)
+)
+
+func isValidIPAddress(host string) bool {
+	return net.ParseIP(host) != nil
+}
+
+func isValidDNSName(host string) bool {
+	if len(host) > 256 {
+		return false
+	}
+
+	labels := strings.Split(host, ".")
+	for _, label := range labels {
+		if !isValidDNSLabel(label) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isValidDNSLabel(label string) bool {
+	if len(label) > 64 {
+		return false
+	}
+
+	return dnsLabelMatcher.Match([]byte(label))
+}
+
+func isValidUnixSocketPath(host string) bool {
+	if !strings.HasPrefix(host, "unix://") || strings.Contains(host, "\000") {
+		return false
+	}
+
+	return true
+}
+
+func validateWorkloadHost(host string) error {
+	// Check that the host is empty
+	if host == "" {
+		return ErrInvalidWorkloadHostFormat{Host: host}
+	}
+
+	// Check if the host represents an IP address, unix socket path or a DNS name
+	if !isValidIPAddress(host) && !isValidUnixSocketPath(host) && !isValidDNSName(host) {
+		return ErrInvalidWorkloadHostFormat{Host: host}
+	}
+
+	return nil
+}
+
+func validateSelector(sel *pbcatalog.WorkloadSelector, allowEmpty bool) error {
+	if sel == nil {
+		if allowEmpty {
+			return nil
+		}
+
+		return errEmpty
+	}
+
+	if len(sel.Names) == 0 && len(sel.Prefixes) == 0 {
+		if allowEmpty {
+			return nil
+		}
+
+		return errEmpty
+	}
+
+	var err error
+
+	// Validate that all the exact match names are non-empty. This is
+	// mostly for the sake of not admitting values that should always
+	// be meaningless and never actually cause selection of a workload.
+	// This is because workloads must have non-empty names.
+	for idx, name := range sel.Names {
+		if name == "" {
+			err = multierror.Append(err, ErrInvalidListElement{
+				Name:    "names",
+				Index:   idx,
+				Wrapped: errEmpty,
+			})
+		}
+	}
+
+	return err
+}
+
+func validateIPAddress(ip string) error {
+	if ip == "" {
+		return errEmpty
+	}
+
+	if !isValidIPAddress(ip) {
+		return errNotIPAddress
+	}
+
+	return nil
+}
+
+func validatePortName(name string) error {
+	if name == "" {
+		return errEmpty
+	}
+
+	if !isValidDNSLabel(name) {
+		return errNotDNSLabel
+	}
+
+	return nil
+}
+
+// generics here are so that the ports map can have either the WorkloadPort or EndpointPort
```

Comment:
I thought Endpoint is using the `WorkloadPort` type so we probably don't need generics anymore.

Reply:
We don't need it but the tests are using the generic behavior. It also makes it really explicit that the WorkloadPort isn't referenced. I am kind of inclined to keep it just to keep it clear from the signature that the map values don't matter at all.

Reply:
Sounds good.